plugins/crypto/ua_securitypolicy_filestore.c

```diff
@@ -0,0 +1,508 @@
+    UA_ByteString localCertificateTmp;
+    UA_ByteString_copy(&securityPolicy->localCertificate, &localCertificateTmp);
+
+    SecurityPolicy_FilestoreContext *pc =
+        (SecurityPolicy_FilestoreContext *) securityPolicy->policyContext;
+
+    UA_StatusCode retval =
+        pc->policy->updateCertificateAndPrivateKey(securityPolicy, newCertificate, newPrivateKey);
+    if(retval != UA_STATUSCODE_GOOD) {
+        return retval;
+    }
+
+    retval =
+        writeCertificateAndPrivateKeyToFilestore(pc->storePath, newCertificate, newPrivateKey);
+    UA_ByteString_clear(&localCertificateTmp);
+
+    return retval;
+}
+
+static UA_StatusCode
+createSigningRequest_sp_filestore(UA_SecurityPolicy *securityPolicy,
+                                  const UA_String *subjectName,
+                                  const UA_ByteString *nonce,
+                                  const UA_KeyValueMap *params,
+                                  UA_ByteString *csr,
+                                  UA_ByteString *newPrivateKey) {
+    SecurityPolicy_FilestoreContext *pc =
+        (SecurityPolicy_FilestoreContext *) securityPolicy->policyContext;
+    return pc->policy->createSigningRequest(pc->policy, subjectName, nonce, params, csr, newPrivateKey);
+}
+
+static void
+clear_sp_filestore(UA_SecurityPolicy *securityPolicy) {
+    if(!securityPolicy)
+        return;
+
+    if(!securityPolicy->policyContext)
+        return;
+
+    SecurityPolicy_FilestoreContext *pc =
+            (SecurityPolicy_FilestoreContext *) securityPolicy->policyContext;
+
+    pc->policy->clear(pc->policy);
+    UA_String_clear(&pc->storePath);
+
+    UA_free(pc->policy);
+    UA_free(pc);
+    securityPolicy->policyContext = NULL;
+}
+
+
+static void
+init_sp_filestore(UA_SecurityPolicy *policy) {
+    SecurityPolicy_FilestoreContext *pc =
+            (SecurityPolicy_FilestoreContext *) policy->policyContext;
+
+    policy->policyUri = pc->policy->policyUri;
+    policy->securityLevel = pc->policy->securityLevel;
+    policy->localCertificate = pc->policy->localCertificate;
+    policy->certificateGroupId = pc->policy->certificateGroupId;
+    policy->certificateTypeId = pc->policy->certificateTypeId;
+    policy->logger = pc->policy->logger;
+
+    UA_SecurityPolicyAsymmetricModule *const asymmetricModule = &policy->asymmetricModule;
+    memcpy(asymmetricModule, &pc->policy->asymmetricModule, sizeof(UA_SecurityPolicyAsymmetricModule));
+    asymmetricModule->makeCertificateThumbprint = asym_makeThumbprint_sp_filestore;
+    asymmetricModule->compareCertificateThumbprint = asym_compareCertificateThumbprint_sp_filestore;
+
+    UA_SecurityPolicySymmetricModule *const symmetricModule = &policy->symmetricModule;
+    memcpy(symmetricModule, &pc->policy->symmetricModule, sizeof(UA_SecurityPolicySymmetricModule));
+    symmetricModule->generateKey = sym_generateKey_sp_filestore;
+    symmetricModule->generateNonce = sym_generateNonce_sp_filestore;
```

Comment:
No offense meant, but this approach of using "trampoline" functions may end up as a maintainer's nightmare. If any crypto-plugin backend function apart from these two decides to use the "securityPolicy->context" in future, you will have to provide an additional trampoline function as well, or we will again run into the same "assume wrong context type" issue. This is hard to track as the implementer of crypto-plugin functions may not be aware of these "wrap-around" SecurityPolicies, and additionally it leads to code bloat since we need to implement trampoline functions for _all_ SecurityPolicies that will support the Certificate File Store, and additionally provide a trampoline function _for each function_ encapsulated inside a SecurityPolicy that uses the securityPolicy->context.
Sorry for ranting - just my two cents...

Reply:
This is now very generic.
The "file-based" security policy is a wrapper that can wrap around any other SecurityPolicy.
So we do not have to duplicate functionality.